Upload migration files from profile settings and start processing

tclaus opened the pull request:
This is a new PR which is rebased on latest dev and a squash of the fixing commits.

tclaus committed:
Upload migration files from profile settings and start processing

This is a rebase on latest dev and a squash of the fixing commits

Flaburgan commented on config/locales/diaspora/en.yml:
```diff
@@ -1318,13 +1317,22 @@ en:
+               To do this, first export your data and photos from the pod you want to move from. Then, import those data and photo archives into this account.
+               Your old account on the other pod will be deleted once this process has completed."
+        warning: "There is no way back; that account can’t be restored if you change your mind."
+        account: "Import another account..."
```
I think we should remove the `...` here.

Flaburgan commented on config/locales/diaspora/en.yml:
```diff
@@ -1339,6 +1347,28 @@ en:
+      list2: "Your settings (theme, default post visibility, etc.) will be transferred from the imported account"
+      list3: "All your aspects and contacts from that account will be added to this account"
+      list4: "All posts and comments from that account will be added to this account"
+      closing_text: "The imported account and all its linked data will be deleted from the previous pod."
```
I guess an archive can be imported only once? Or what would happen if I import the same archive to two pods at the same time? The old pod will answer to the second import as if the user does not exist I assume? Maybe we should mention it here?

tclaus replied:
What do you recommend? "Be careful", "After the import you can't start the process another time" ? Any ideas?

Flaburgan commented on app/views/users/_edit.haml:
```diff
@@ -229,7 +233,20 @@
+          %strong
+            = t(".import.warning")
+        .form-group
+          .btn.btn-primary{id: "import_account", data: {toggle: "modal", target: "#importAccountModal"}}
```
I think after validating the modal and while the import is ongoing, this blue button "import another account" should be removed and replaced by a sentence "Account xxx@ccc is being imported, please come back in a few minutes" like we are doing for the export.

tclaus replied:
Button is now invisible for users with a pending / ongoing import.

Flaburgan reviewed (changes requested):
Hey, so I just tried this locally, and it worked great, although there were no comments in my posts, apart from mine. Is this because I did the import offline, and they are supposed to be pulled? But then if the post is private or the old pod offline, this won't work. I guess it's "normal" to not include in the archive the comments made by others (although we could discuss that I guess), but then I think we should inform the user.
I also made a few suggestions to improve the UX. I'm going to read the code now.

tclaus committed:
Style fixes. Remove import button while import is in progress

Flaburgan commented:
Hello!
The button being disabled does not work locally, it's because I don't have sidekiq there right?

Can you also enable the message on the registration page by removing this if? https://github.com/diaspora/diaspora/blob/develop/app/views/registrations/_new.erb#L57

Thanks!

tclaus committed:
Show migration info on registration page

tclaus commented:
@Flaburgan It should work, when a migration is completed. (This worked locally) BTW I have disabled the If to show the message on registration page.

Flaburgan commented:
I just deployed this branch to diaspora-fr and tried to import my diasp.org account but it did not work. I submitted the dialog as I did locally, but the Import button is still available in my settings. I  don't know if the jobs was actually scheduled in sidekiq, I just realized that despite setting `log = log/sidekiq.log` in diaspora.toml, there actually is no log file for sidekiq there, only the production one.

Flaburgan commented:
Another remark: after the import, I have been spammed by email notifications of all my imported contacts, that they have started sharing with me. I think we don't need to have a notification about this, as the user is triggering it.

tclaus commented:
> I just deployed this branch to diaspora-fr and tried to import my diasp.org account but it did not work. I submitted the dialog as I did locally, and the green message "Import is scheduled" but the Import button is still available in my settings. I don't know if the jobs was actually scheduled in sidekiq, I just realized that despite setting `log = log/sidekiq.log` in diaspora.toml, there actually is no log file for sidekiq there, only the production one.
> 
> Edit: maybe no sidekiq log file is due to the fact that I start it with systemd. I checked them with journalctl but no mention of `import` there.
> 
For the records: $: `sudo journalctl -u diaspora-sidekiq` or `sudo journalctl -u sidekiq` shows the logs

Flaburgan commented:
@tclaus any chance you can have a look at the "Import" button not disappearing when closing th dialog? I wish I can tell people to come to diaspora-fr.org before diasp.org is closed.

tclaus commented:
To disable the button I have made some deeper investigation - first, the dialog is submitted - here Javascript may do the job to disable the button immediately,
Import file is uploaded and validation stars - here an enum (new column) in the user-table may indicate an ongoing validation check (status: “Import Validation”)
Then a job is triggering (status: “Import job pending”)
Then the job could set a “Import running” and at least “import completed” or “import failed”.

I think on a new column, with some statuses, to indicate that an import has been started and its current status.
IMHO this kind of status does not exist currently?

What did you think? Is this a way we can go?

Flaburgan commented:
>first, the dialog is submitted - here Javascript may do the job to disable the button immediately,

I think when you submit the form you should do a POST redirect GET. At the moment if the users press F5 after the dialog is closed, it is re-submitting the form, that should not happen. To reload the page entirely will avoid this problem, and allow us to display the correct status instead of the button.

Then about how to get this information, how is this done for the export buttons? Because everything is working nicely for those, so I think we should do the same.

Flaburgan commented:
So I just checked how it is done for exporting, and indeed we have two booleans on the user table, one [exporting](https://github.com/diaspora/diaspora/blob/develop/db/migrate/0000_create_schema.rb#L613) for the data and one [exporting_photos](https://github.com/diaspora/diaspora/blob/develop/db/migrate/0000_create_schema.rb#L617) for the pictures.
They are then [set to true when the export starts](https://github.com/diaspora/diaspora/blob/develop/app/models/user.rb#L315) and [back to false](https://github.com/diaspora/diaspora/blob/develop/app/models/user.rb#L325) when it's done. So I guess we should do the same for the import.
Will you have time to have a look in the coming days, or should I try (even if I never modified the schema or anything similar).

tclaus commented:
@Flaburgan 
> So I just checked how it is done for exporting, and indeed we have two booleans on the user table, one [exporting](https://github.com/diaspora/diaspora/blob/develop/db/migrate/0000_create_schema.rb#L613) for the data and one [exporting_photos](https://github.com/diaspora/diaspora/blob/develop/db/migrate/0000_create_schema.rb#L617) for the pictures. They are then [set to true when the export starts](https://github.com/diaspora/diaspora/blob/develop/app/models/user.rb#L315) and [back to false](https://github.com/diaspora/diaspora/blob/develop/app/models/user.rb#L325) when it's done. So I guess we should do the same for the import. Will you have time to have a look in the coming days, or should I try (even if I never modified the schema or anything similar).

Sounds OK for me - let's include one boolean for importing user data and one for photos. The details are already stored in the table 'account_migrations'. 

This evening I have some time to implement this.

tclaus committed:
Upload and close the modal dialog.

Sets an importing flag to the user account.

Flaburgan commented on db/migrate/20250327181326_add_importing_data_field_to_user.rb:
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class AddImportingDataFieldToUser < ActiveRecord::Migration[6.1]
+  def change
```
This is failing:
```
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

wrong number of arguments (given 3, expected 2)
/home/andupara/.rvm/gems/ruby-3.1.4/gems/activerecord-6.1.7.8/lib/active_record/connection_adapters/abstract/schema_definitions.rb:603:in `column'
/home/andupara/diaspora/diaspora/db/migrate/20250327181326_add_importing_data_field_to_user.rb:6:in `block in change'
```

tclaus committed:
Fixing syntax error in mogration

tclaus committed:
Styles

Flaburgan commented:
@tclaus hey, it looks like with your last commits the import is not launched anymore: I don't have the green confirmation popping up, and it does not look like it starts. Why did you do the POST with JS, I think it was working fine the way it was before, but in the rails controller at the end you should simply redirect to the user settings, instead of returning it directly.

tclaus committed:
Re-enabling flash messages

tclaus committed:
Clear new fields on clear_account

tclaus commented:
> @tclaus hey, it looks like with your last commits the import is not launched anymore: I don't have the green confirmation popping up, and it does not look like it starts. Why did you do the POST with JS, I think it was working fine the way it was before, but in the rails controller at the end you should simply redirect to the user settings, instead of returning it directly.

Hi @Flaburgan , I thought the POST with JS in combinati0on was the only way to hide the dialog after uploading, but I was wrong. 
Changed this back and enabled the flash-message again. (Was a simple parameter in the flash statement that prevented the flash in the UI - I fixed this. 
I also fixed another failing test.

Also without a visible flag message the import should have been startet (I can see in the debugger a start of the import job every time)
